padrino-core/lib/padrino-core/path_router/matcher.rb

```diff
@@ -38,9 +38,35 @@ def to_regexp
       #
       def expand(params)
         params = params.merge(@default_values) if @default_values.is_a?(Hash)
-        expanded_path = handler.expand(:append, params)
+        params, query = params.each_with_object([{}, {}]) do |(key, val), parts|
+          parts[handler.names.include?(key.to_s) ? 0 : 1][key] = val
+        end
+        query = build_nested_query(query)
+        expanded_path = Rack::Utils.unescape(handler.expand(:append, params))
+        expanded_path += ?? + query unless query.empty?
         expanded_path
       end
+
+      ##
+      # Builds nested query.
+      #
+      def build_nested_query(value, prefix = nil)
+        case value
+        when Array
+          value.map { |v|
+            build_nested_query(v, "#{prefix}[]")
+          }.join("&")
+        when Hash
+          value.map { |k, v|
+            build_nested_query(v, prefix ? "#{prefix}[#{k}]" : k)
+          }.reject(&:empty?).join('&')
+        when nil
+          prefix
```

Comment:
Should be `prefix.to_s`, no? If `prefix` is nil it returns nil which does not look like a valid query.